cpp/src/io/parquet/chunk_dict.cu

```diff
@@ -0,0 +1,379 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <io/parquet/parquet_gpu.hpp>
+
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/table/row_operators.cuh>
+#include "cudf/detail/iterator.cuh"
+#include "rmm/exec_policy.hpp"
+
+namespace cudf {
+namespace io {
+namespace parquet {
+namespace gpu {
+
+template <int block_size>
+__global__ void __launch_bounds__(block_size, 1)
+  gpuInitializeChunkHashMaps(device_span<EncColumnChunk> chunks)
+{
+  auto chunk = chunks[blockIdx.x];
+  auto t     = threadIdx.x;
+  // TODO: Now that per-chunk dict is same size as ck.num_values, try to not use one block per chunk
+  for (size_t i = 0; i < chunk.dict_map_size; i += block_size) {
+    if (t + i < chunk.dict_map_size) {
+      new (&chunk.dict_map_slots[t + i].first) map_type::atomic_key_type{KEY_SENTINEL};
+      new (&chunk.dict_map_slots[t + i].second) map_type::atomic_mapped_type{VALUE_SENTINEL};
```

Comment:
This is placement new. If somebody is not familiar with this, they may misunderstand this is something else that can cause memory leak. Please add comments to explain it.